File: ConnectFour.java
import java.util.Scanner;
public class ConnectFour{
	public static void main(String[] args){
		String[][] board = fillBoard();
		game(board);

	}
	public static String[][] fillBoard(){//fills board empty spaces (or else command line prints out a bunch of null)
		String[][] arr = new String[6][7];
		for(int i = 0; i < arr.length; i++){
			for(int j = 0; j < arr[0].length; j++){
					arr[i][j] = " ";
			}
		}
		return arr;
	}
	public static void printArr(String[][] arr){//prints the board in a stylized manner
		System.out.println("\n 1   2   3   4   5   6   7\n---------------------------");
		for(int i = 0; i < arr.length; i++){
			for (int j = 0; j < arr[0].length; j++){
				System.out.print(" " + arr[i][j] + " |");
			}
			System.out.println("\n---------------------------");
		}
	}
	public static void game(String[][] board){
		boolean player1 = true;
		Scanner sc = new Scanner(System.in);
		String s;
		int turn = 1;
		int[] checkers = {0,0,0,0,0,0,0}; //keeps track of how high the checkers are stacked in a column
		int moveP;

		while(turn!= 43){
			if(player1){
				s = "O";
			}else{
				s = "X";
			}
			printArr(board);
			while(true){
				int move = sc.nextInt();
				move--;
				if(move > 6 || move < 0 || checkers[move] >= 6){
					System.out.println("invalid move try again");
					continue;
				}
				moveP = move;
				board = placeMove(board, s, move, checkers);
				break;
			}
			if(victoryCheck(board, s,checkers, moveP)){
				break;
			}
			player1 = !player1; //switches player to opposite condition
			turn++;
		}
		printArr(board);
		if(turn == 43)
			System.out.println("player 1 and 2 tied");
		else{
			int win = 1; //used to refer to player 1 or 2
			if(!player1)
				win++;
			System.out.println("Player " + win + " won!!!");
		}
		System.exit(0);
	}
	public static String[][] placeMove(String[][] arr, String s, int move, int[] checkers){
		arr[arr.length-checkers[move]-1][move] = s;
		checkers[move]++;
		return arr;
	}
	public static boolean victoryCheck(String[][] arr, String s, int[] checkers, int move){//checks to see if a player won
		boolean[] pause = new boolean[8];
		int height = arr.length-1;
		int length = arr[0].length-1;
		int i = height - checkers[move] + 1; //gets vertical position of move
		int row = 1, col = 1, diagR = 1, diagL = 1;
			for(int j = 1; j < arr[0].length; j++){
				if(pause[0] && pause[1] && pause[2] && pause[3] && pause[4] && pause[5] && pause[6] && pause[7])
					break;
				System.out.println(j);

			//counts rows
				if(!pause[0] && move + j <= length && arr[i][move + j].equals(s)){ //to right
					row++;
				}else{
					pause[0] = true;//stops counting once streak ends
				}
				if(!pause[1] && move - j >= 0 && arr[i][move - j].equals(s)){ //to left
					row++;
				}else{
					pause[1] = true;
				}
			//counts columns
				if(!pause[2] && i + j <= height && arr[i + j][move].equals(s)){ //up
					col++;
				}else{
					pause[2] = true;
				}
				if(!pause[3] && i - j >= 0 && arr[i - j][move].equals(s)){ //down
					col++;
				}else{
					pause[3] = true;
				}
			//counts diagonals
				if(!pause[4] && move + j <= length && i + j <= height && arr[i + j][move + j].equals(s)){ //to top right
					diagR++;
				}else{
					pause[4] = true;
				}
				if(!pause[5] && i - j >= 0 && move - j >= 0 && arr[i-j][move-j].equals(s)){ //to bottom left
					diagL++;
				}else{
					pause[5] = true;
				}
				if(!pause[6] && move - j >= 0 && i + j <= height && arr[i+j][move-j].equals(s)){ //to top left
					diagL++;
				}else{
					pause[6] = true;
				}
				if(!pause[7] && move + j <= length && i - j >= 0 && arr[i - j][move + j].equals(s)){ //to bottom right
					diagR++;
				}else{
					pause[7] = true;
				}
			}
			if(row >= 4 || col >= 4 || diagR >= 4 || diagL >= 4)//this needs to be inside nested loop
				return true;
		return false;

	}
/*
	public static boolean victoryCheck(String[][] arr, String s){//checks to see if a player won
		//check row/collumn victory
			for(int i = 0; i < arr.length; i++){
				int countRow = 0;
				int countCol = 0;
				for(int j = 0; j < arr[0].length; j++){
					if(arr[i][j].equals(s))
						countRow++;
					else
						countRow = 0;//resets if opponent piece blocks a chain
					if(j < 6 && arr[j][i].equals(s))
						countCol++;
					else
						countCol = 0;
					if(countRow >= 4 || countCol >=4)
						return true;
				}
			}
		//check diagonal victory
			for(int i = 0; i < 2; i++){
				int countD1 = 0, countD2 = 0, countD3 = 0, countD4 = 0;//resets diagonal count
				for(int j = 0; j < 5 - i; j++){
					
					if(arr[5-j][1+j].equals(s))//counts from bottom left to top right
						countD1++;
					else
						countD1 = 0;
					if(arr[5-j][5-j].equals(s))//counts from bottom right to top left
						countD2++;
					else
						countD2 = 0;
					if(arr[0+j][1+j].equals(s))//counts from top left to bottom right
						countD3++;
					else
						countD3 = 0;
					if(arr[0+j][5-j].equals(s))//counts from top right to bottom left
						countD4++;
					else
						countD4 = 0;
					if(countD1 == 4 || countD2 == 4 || countD3 == 4 || countD4 == 4)//this needs to be inside nested loop
						return true;
				}
			}
		return false;

	}
*/
}

File: ConnectFour.js
function fillBoard() {//fills array with blank spaces
    let boardArr = new Array(6);
    for (let i = 0; i < boardArr.length; i++) {
        boardArr[i] = new Array(7).fill(" ");
    }
    return boardArr;
}

const printBoard = (board) => { //print board for console
    console.log("\n 1   2   3   4   5   6   7\n---------------------------");
    for(let i = 0; i < board.length; i++){
        let row = "";
        for (let j = 0; j < board[0].length; j++){
            row += " " + board[i][j] + " |";
        }
        console.log(row);
        console.log("---------------------------");
    }
}

const placeMove = (board, symbol, move, checkers) => {
	board[board.length-checkers[move] - 1][move] = symbol;
	checkers[move]++;
	return board;
}

const victoryCheck = (board, symbol, checkers, move) => { 
    let c1=false,c2=false,r=false,dR1=false,dR2=false,dL1=false,dL2=false
    let height = board.length-1;
    let length = board[0].length-1;
    let i = height - checkers[move]+1;

    let row = 1, column = 1, diagonalRight = 1, diagonalLeft = 1;
    for(let j = 1; j < 4 && !(c1&&c2&&r&&dR1&&dR2&&dL1&&dL2); j++){

        //counts columns
        if(!c1 && move + j <= length && board[i][move + j] === symbol){ //to right
            column++;
        }else{
            c1 = true;//stops counting once streak ends
        }
        if(!c2 && move - j >= 0 && board[i][move - j] === symbol){ //to left
            column++;
        }else{
            c2 = true;
        }

        //counts rows

        if(!r && i + j <= height && board[i + j][move] === symbol){ //down
            row++;
        }else{
            r = true;
        }

        //counts diagonals
        if(!dR1 && move + j <= length && i + j <= height && board[i + j][move + j] === symbol){ //to top right
            diagonalRight++;
        }else{
            dR1 = true;
        }
        if(!dR2 && move - j >= 0 && i - j >= 0 && board[i-j][move-j] === symbol){ //to bottom left
            diagonalRight++;
        }else{
            dR2 = true;
        }
        if(!dL1 && move - j >= 0 && i + j <= height && board[i+j][move-j] === symbol){ //to top left
            diagonalLeft++;
        }else{
            dL1 = true;
        }
        if(!dL2 && move + j <= length && i - j >= 0 && board[i - j][move + j] === symbol){ //to bottom right
            diagonalLeft++;
        }else{
            dL2 = true;
        }
    }
    console.log("Row: " + row + "\nColumn: " + column + "\nRight Diagonal: " + diagonalRight + "\nLeft Diagonal: " + diagonalLeft);//used for debugging purposes
    return (row >= 4 || column >= 4 || diagonalRight >= 4 || diagonalLeft >= 4);
}

const game = (board) => {
	let player1 = true;
	let turn = 1;
	const checkers = [0,0,0,0,0,0,0]; //keeps track of how high the checkers are stacked in a column
	let move;
	const prompt = require('prompt-sync')();
	while(turn != 43){ //game loop
		let symbol = player1 ? 'O' : 'X'; //if player1 is(?) true(O):false(X) (ternary operator)
		printBoard(board);

		while (true) {
		  move = prompt("Choose a column you would like to make a move in");
		  move = parseInt(move)-1; // subtract 1 to make it zero-indexed
		  if (move > 6 || move < 0 || checkers[move] >= 6) {
		    console.log("Invalid move, please try again");
		  } else {
		    board = placeMove(board, symbol, move, checkers);
		    break;
		  }
		}
		
		if(victoryCheck(board, symbol, checkers, move)){
			break;
		}
		player1 = !player1; //switches player to opposite condition
		turn++;
	}

	printBoard(board);

	if(turn == 43)
		console.log("Player 1 and Player 2 tied");
	else {
		let win = 1; //used to refer to player 1 or 2
		if(!player1)
			win++;
		console.log("Player " + win + " won!!!");
	}
	process.exit(0);
}
let board = fillBoard(); //board is a 2D array
game(board) //start the game
